src/render.rs

```diff
@@ -345,6 +363,71 @@ impl<'a> Renderer<'a> {
         Ok(output.trim_right().to_string())
     }
 
+    // TODO: have a look at this method once it is working to clean it up, too much indentation
+    // Probably very inefficient and can be trimmed down
+    fn render_macro(&mut self, call_node: Node) -> TeraResult<String> {
+        match call_node {
+            MacroCall {namespace, name, params} => {
+                // Avoid shadowing when matching the definition of macros
+                let call_params = params;
+                // To find the real namespace we just look at the last one we got.
+                // If we have self, we take the last one otherwise we push it into the vec
+                let real_namespace = if namespace == "self" {
+                    // TODO: handle error if we don't have a namespace
+                    self.macro_namespaces
+                        .last()
+                        .expect("Open an issue with a template sample please (mention `self namespace macro`)!")
+                        .to_string()
+                } else {
+                    self.macro_namespaces.push(namespace.clone());
+                    namespace.clone()
+                };
+                // TODO: avoid cloned() everywhere if possible
+                match self.macros.get(&real_namespace).cloned() {
```

Comment:
This is very, very hairy, and likely needlessly so. Use the `Option` combinators to make it much more accessible: `map`, `and_then`, etc.